crates/server/src/server/server_future.rs

```diff
@@ -642,71 +673,146 @@ impl<T: RequestHandler> ServerFuture<T> {
+    /// new connections and the returned future will complete once all tasks have terminated.
+    ///
+    /// This is equivalent to calling [Drain::drain] followed by [Self::block_until_done].
+    pub fn shutdown_gracefully(self) -> impl Future<Output = Result<(), ProtoError>> {
+        self.drain.drain();
+        self.block_until_done()
+    }
+
+    /// This will run until all background tasks complete. If one or more tasks return an error,
+    /// one will be chosen as the returned error for this future.
     pub async fn block_until_done(mut self) -> Result<(), ProtoError> {
-        let result = self.join_set.join_next().await;
+        if self.join_set.is_empty() {
+            warn!("block_until_done called with no pending tasks");
+            return Ok(());
+        }
 
-        match result {
-            None => {
-                tracing::warn!("block_until_done called with no pending tasks");
-                Ok(())
+        // Now wait for all of the tasks to complete.
+        let mut out = Ok(());
+        while let Some(join_result) = self.join_set.join_next().await {
+            match join_result {
+                Ok(result) => {
+                    match result {
+                        Ok(_) => (),
+                        Err(e) => {
+                            // Save the last error.
+                            out = Err(e);
+                        }
+                    }
+                }
+                Err(e) => return Err(ProtoError::from(format!("Internal error in spawn: {e}"))),
             }
-            Some(Ok(x)) => x,
-            Some(Err(e)) => Err(ProtoError::from(format!("Internal error in spawn: {e}"))),
+        }
+        out
+    }
+}
+/// A Future that is used to signal server threads to begin a graceful shutdown by
+/// draining connections/requests.
+#[derive(Clone)]
+pub struct Drain {
+    drain: Arc<AtomicBool>,
+}
+
+impl Drain {
+    /// Create a new server handle.
+    pub fn new() -> Self {
+        Self {
+            drain: Arc::new(AtomicBool::new(false)),
+        }
+    }
+
+    /// Signals server threads to begin a graceful shutdown by draining connections/requests
+    pub fn drain(&self) {
+        self.drain.store(true, std::sync::atomic::Ordering::Relaxed);
+    }
+
+    /// Indicates whether a graceful shutdown has been initiated.
+    pub fn is_draining(&self) -> bool {
+        self.drain.load(std::sync::atomic::Ordering::Relaxed)
+    }
+}
+
+impl Future for Drain {
+    type Output = ();
+
+    fn poll(self: Pin<&mut Self>, _: &mut Context<'_>) -> Poll<Self::Output> {
+        if self.is_draining() {
+            Poll::Ready(())
+        } else {
+            Poll::Pending
         }
     }
 }
 
 /// Reap finished tasks from a `JoinSet`, without awaiting or blocking.
 fn reap_tasks(join_set: &mut JoinSet<()>) {
-    while FutureExt::now_or_never(join_set.join_next()).is_some() {}
+    if !join_set.is_empty() {
```

Comment:
Why is this necessary?

Reply:
I moved where this is called to the outside of the loops, where there may or may not be tasks in the join_set. If `FutureExt::now_or_never` is called with None, it hangs forever.